lib/srv/db/mongodb/autousers.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mongodb
+
+import (
+	"cmp"
+	"context"
+	"slices"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+type user struct {
+	Username         string                `bson:"user"`
+	Roles            userRoles             `bson:"roles"`
+	CustomData       userCustomData        `bson:"customData"`
+	AuthRestrictions []userAuthRestriction `bson:"authenticationRestrictions"`
+}
+
+func (u *user) isLocked() bool {
+	for _, authRestriction := range u.AuthRestrictions {
+		if slices.Contains(authRestriction.ClientSource, lockedClientSource) {
+			return true
+		}
+	}
+	return false
+}
+
+type userCustomData struct {
+	TeleportAutoUser bool `bson:"teleport-auto-user"`
+}
+
+type userAuthRestriction struct {
+	ClientSource []string `bson:"clientSource"`
+}
+
+type userRole struct {
+	Rolename string `bson:"role"`
+	Database string `bson:"db"`
+}
+
+type userRoles []userRole
+
+func (r userRoles) sort() {
+	slices.SortFunc(r, func(a, b userRole) int {
+		if cmpDatabase := cmp.Compare(a.Database, b.Database); cmpDatabase != 0 {
```

Comment:
nit:  Does string comparison < won't work in this case ?

Reply:
the compare function expects an `int` return, so it's easier to use `cmp.Compare` in my opinion and use it for return in next line.